lib/plugin_manager.rb

```diff
@@ -0,0 +1,205 @@
+      File.read(gemfile_plugins_path)
+    end
+  end
+
+  def add(name)
+    plugin = Plugin.new(name)
+    if plugin.included_in?(@gemfile_plugins)
+      puts 'Plugin already installed, abort'
+      exit
+    end
+    _write_plugin_to_gemfile_plugins_file(plugin)
+    _bundle
+    _migrate
+
+    if @environment == 'production'
+      _assets_precompile
+    else
+      _assets_webpack
+    end
+  end
+
+  def remove(name)
+    plugin = Plugin.new(name)
+    unless plugin.included_in?(@gemfile_plugins)
+      puts 'Plugin not installed, abort!'
+      exit
+    end
+    _revert_migrations(plugin)
+    _remove_plugin_from_gemfile_plugins_file(plugin)
+    _bundle
+    if @environment == 'production'
+      _assets_clobber
+      _assets_precompile
+    end
+  end
+
+  def _write_plugin_to_gemfile_plugins_file(plugin)
+    _add_plugin_to_gemfile_plugins(plugin)
+    _sort_gemfile_plugins
+    _write_to_gemfile_plugins_file
+  end
+
+  def _add_plugin_to_gemfile_plugins(plugin)
+    @gemfile_plugins << plugin.gemfile_plugins_line
+    @gemfile_plugins << plugin.gemfile_plugins_lines_for_dependencies
+  end
+
+  def _sort_gemfile_plugins
+    # todo we have to take of the order..
+    # e.g. global roles first and
+    # reporting engine before costs/reporting
+  end
+
+  def _write_to_gemfile_plugins_file
+    gemfile_plugins_path = 'Gemfile.plugins'
+    File.open(gemfile_plugins_path, 'w') do |f|
+      f.write @gemfile_plugins
+    end
+  end
+
+  def _bundle
+    Bundler.with_clean_env do
+      system 'bundle install --no-deployment'
+    end
+  end
+
+  def _migrate
+    # todo should we migrate for all envs?
+    system "RAILS_ENV='#{@environment}' rake db:migrate"
+  end
+
+  def _revert_migrations(plugin)
+    # todo should we migrate for all envs?
+    migration_path = OpenProject::Application.config.paths['db/migrate'].select { |path| path.include?(plugin.name) }
+    ActiveRecord::Migrator.migrate migration_path, 0
+  end
+
+  def _assets_precompile
+    system "RAILS_ENV='production' rake assets:precompile"
+  end
+
+  def _assets_webpack
+    system "RAILS_ENV='production' rake assets:webpack"
+  end
+
+  def _assets_clobber
+    system "RAILS_ENV='production' rake assets:clobber"
+  end
+
+  def _remove_plugin_from_gemfile_plugins_file(plugin)
+    _remove_plugin_from_gemfile_plugins(plugin)
+    _write_to_gemfile_plugins_file
+    _remove_gemfile_plugins_if_empty
+  end
+
+  def _remove_plugin_from_gemfile_plugins(plugin)
+    gemfile_plugins_new = ''
+    plugins = [plugin].concat _dependencies_only_for(plugin)
+    @gemfile_plugins.each_line do |line|
+      gemfile_plugins_new << line if _line_contains_no_plugin?(line, plugins)
+    end
+    @gemfile_plugins = gemfile_plugins_new
+  end
+
+  def _dependencies_only_for(plugin)
+    # todo this needs to be addressed
+    dependencies = plugin.dependencies
+    dependencies.select { |dependency| true }#todo dependency._not_needed_by_any_other_than?(plugin) }
```

Comment:
Unused block argument - `dependency`. You can omit the argument if you don't care about it.<br>Line is too long. [102/99]<br>Missing space after #.<br>Put a space before an end-of-line comment.